File: R/xml2tps.r
library(XML)
library(stringr)
library(data.table)

xml2tps <- function(xmlfile, tpsfile, fname = "", LM = 0, top = 0) {
  
  xml_data <- xmlParse(xmlfile)
  
  images <- getNodeSet(xml_data, "//image")
  
  csvfile <- "output.csv"
  csv_con <- file(csvfile, open = "wt")
  
  header <- c("id")
  for (i in 0:(LM - 1)) {
    header <- c(header, paste0("X", i), paste0("Y", i))
  }
  writeLines(paste(header, collapse = ","), csv_con)
  
  img_id_counter <- 0
  
  for (image in images) {
    img_file <- xmlGetAttr(image, "file")
    img_file <- str_replace(img_file, fname, "")
    
    boxes <- getNodeSet(image, "box")
    
    for (box in boxes) {
      row <- c(img_file)
      
      parts <- getNodeSet(box, "part")
      for (part in parts) {
        x <- xmlGetAttr(part, "x")
        y <- as.numeric(xmlGetAttr(part, "y"))
        y <- abs(top - y)
        row <- c(row, x, y)
      }
      
      writeLines(paste(row, collapse = ","), csv_con)
      
      img_id_counter <- img_id_counter + 1
    }
  }
  
  close(csv_con)
  
  archivo <- fread(csvfile)
  
  matrix <- as.matrix(archivo)
  nombres <- matrix[, 1]
  coordenadas <- matrix[, -1]
  
  valoresX <- coordenadas[, seq(1, ncol(coordenadas), by = 2)]
  valoresY <- coordenadas[, seq(2, ncol(coordenadas), by = 2)]
  
  con <- file(tpsfile, open = "wt")
  
  for (i in 1:nrow(matrix)) {
    cat(paste("LM=", LM, sep = ""), file = con)
    
    for (j in 1:ncol(valoresX)) {
      cat(paste("\n", valoresX[i, j], valoresY[i, j], sep = " "), file = con)
    }
    
    cat(paste("\nIMAGE=", nombres[i], sep = ""), file = con)
    cat(paste("\nID=", i - 1, sep = ""), file = con)
    
    if (i < nrow(matrix)) {
      cat("\n", file = con)
    }
  }
  
  close(con)
  
  unlink(csvfile)
}

File: Python/xml2tps.py
def xml2tps(xmlfile, tpsfile, images, fname = '', LM = 0):   
    import os
    import pandas as pd 
    import xml.etree.ElementTree as ET
    import csv
    from PIL import Image

    image_sizes = {}
    valid_extensions = ('.jpg', '.jpeg', '.png', '.bmp', '.gif', '.tiff')  # Image's extensions

    for filename in os.listdir(images):
        if filename.lower().endswith(valid_extensions):
            image_path = os.path.join(images, filename)
            with Image.open(image_path) as img:
                image_sizes[filename] = img.size  
    
    alturas = []
    for size in image_sizes.items():
        alturas.append(size[1][1])
    
    tree = ET.parse(xmlfile)
    root = tree.getroot()

    with open("output.csv", "w", newline = "") as csvfile:
        csvwriter = csv.writer(csvfile)
        
        ID = ["id"]
        coordenadas = []
        for i in range(LM):
            valoresx, valoresy = f"X{i}", f"Y{i}"
            coordenadas.append(valoresx)
            coordenadas.append(valoresy)
        valores = ID + coordenadas
        csvwriter.writerow(valores)
    
        for image in root.find('images').findall('image'):
            file_id = image.get('file')
            box = image.find('box')
            parts = box.findall('part')
            coordinates = []
            for part in parts:
                x = part.get('x')
                y = part.get('y')
                coordinates.extend([x, y])

            row = [file_id] + coordinates
            csvwriter.writerow(row)
    
    archivo = pd.read_csv("output.csv")
    
    archivo["id"] = archivo["id"].str.replace(f"{fname}", "")
        
    matrix = archivo.to_numpy()
    nombres = []
    coordenadas = []
    valoresX = matrix[:, 1::2]
    valoresY = matrix[:, 2::2]

    for i in matrix:
        nombres.append(i[0])
    for i in range(len(valoresX)):
        lista_tuplas = list(zip(valoresX[i], abs(alturas[i] - valoresY[i])))
        coordenadas.append(lista_tuplas)
    data = {'id': nombres, 'landmarks': coordenadas}

    with open(tpsfile, 'w') as file:
        contador = 0
        for idx, id in enumerate(data['id']):
            file.write(f"LM={int(len(valoresX[0]))}\n")
            for (x, y) in data['landmarks'][idx]:
                file.write(f"{x} {y}\n")
            file.write(f"IMAGE={id}\n")
            file.write(f"ID={contador}\n")
            contador += 1
    with open(tpsfile, 'w') as file:
        contador = 0
        for idx, id in enumerate(data['id']):
            file.write(f"LM={int(len(valoresX[0]))}\n")
            for (x, y) in data['landmarks'][idx]:
                file.write(f"{x} {y}\n")
            file.write(f"IMAGE={id}\n")
            file.write(f"ID={contador}\n")
            contador += 1

    os.remove("output.csv")